src/EditorFeatures/CSharpTest/UseRecursivePatterns/UseRecursivePatternsTests.cs

```diff
@@ -0,0 +1,1077 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.UseRecursivePatterns;
+using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeRefactorings;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.UseRecursivePatterns
+{
+    public class UseRecursivePatternsTests : AbstractCSharpCodeActionTest
+    {
+        protected override CodeRefactoringProvider CreateCodeRefactoringProvider(Workspace workspace, TestParameters parameters)
+           => new CSharpUseRecursivePatternsRefactoringProvider();
+
+        [Fact]
+        public async Task Test00()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    bool M(C b)
+    {
+        return b.P1 == 1 [||]|| b.P1 == 2;
+    }
+    public int P1, P2;
+}",
+@"class C
+{
+    bool M(C b)
+    {
+        return b is { P1: 1 or 2 };
+    }
+    public int P1, P2;
+}");
+        }
+
+        [Fact]
+        public async Task Test01()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    bool M(C b)
+    {
+        return b.P1 == null [||]&& b.P2 == null;
+    }
+    public object P1, P2;
+}",
+@"class C
+{
+    bool M(C b)
+    {
+        return b is { P1: null, P2: null };
+    }
+    public object P1, P2;
+}");
+        }
+
+        [Fact]
+        public async Task Test02()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    bool M(C b)
+    {
+        return b.MetadataName == ""a"" &&
+            b.ContainingType == null &&
+            b.ContainingNamespace != null &&
+            b.ContainingNamespace.Name == ""a"" [||]&&
```

Comment:
will this work if it's written as : `b.ContainingNamespace?.Name == ""a""`?

Reply:
(I'm gonna respond to some comments and then go about it)



Yes. See `VisitInput` which deconstruct a conditional access into a chain of Evaluation nodes. Will add a test and comment.